subsys: Add network core monitor (#11979)
nrfconnect/sdk-nrf: doc/nrf/libraries/others/network_core_monitor.rst

.. _network_core_monitor:

Network core monitor
####################

.. contents::
   :local:
   :depth: 2

This library monitors the network core status of the nRF5340 processor.

Overview
********

It detects suspensions and resets in the network processor.
When a reset occurs, you can read the cause of the reset.

Implementation
==============

The library uses two general-purpose registers of the IPC peripheral in the application core and writes the state of the network processor to them.

.. figure:: images/ncm_register.svg
   :alt: Application of IPC peripheral registers

   IPC peripheral registers

> The ``GPMEM[0]`` register is divided into two 16bit parts.

Reviewer: ```suggestion
The ``GPMEM[0]`` register is divided into two 16-bit parts.
```

> The ``COUNTER`` value is incremented by the network core every :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC`.

Reviewer 2: ```suggestion
The ``COUNTER`` value (in milliseconds) is incremented by the network core with the same value that is originally set in the :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC` Kconfig option.
```
I would add more details here - is the sentence still correct?

Author: `COUNTER` is not in milliseconds. 
This is a counter incremented by 1 every `CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC` millisecond.

Reviewer 2: I see, thanks. If we remove the incorrect part in brackets, would it be sufficient?

Author: After the off-line discussion, this sentence will remain unchanged.

The application core periodically calls the :c:func:`ncm_net_status_check` function to check the status of the network core.
If the network core is suspended, the counter value is not updated.
The :c:func:`ncm_net_status_check` function returns error code ``-EBUSY``.

To correctly detect the network core status, the :c:func:`ncm_net_status_check` function should be called less frequently than the value set in the :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC` Kconfig option.

> The ``FLAGS`` field has the ``Reset`` flag as the bit 0."

Reviewer 2: ```suggestion
The ``FLAGS`` field has the ``Reset`` flag as the bit 0.
```

.. _network_core_monitor:

> Network Core Monitor

Reviewer 3: ```suggestion
Network core monitor
```

####################

.. contents::
   :local:
   :depth: 2

> Network core status monitoring of the nRF5340 processor.

Reviewer 3: ```suggestion
This library monitors the network core status of the nRF5340 processor.
```

Overview
********

> This library monitors the network core status of the nRF5340 processor by application core.

Reviewer 3: Maybe you could leave this line out as it is pretty close to the description before the overview.

> The Network Core Monitor module provides network core status monitoring by application core for the nRF5340 processor.

Reviewer 3: ```suggestion
This library monitors the network core status of the nRF5340 processor by application core.
```

> It allows detection of network processor suspensions and resets.

Reviewer 3: ```suggestion
It detects suspensions and resets in the network processor.
```

> After a reset occurs, it is possible to read the cause of the reset.

Reviewer 3: ```suggestion
When a reset occurs, you can read the cause of the reset.
```

Implementation
==============

> The module uses two general-purpose registers of the IPC peripheral of the application core.
> It writes the state of the network processor to them.

Reviewer 3: ```suggestion
The library uses two general-purpose registers of the IPC peripheral in the application core and writes the state of the network processor to them.
```

> .. figure:: images/ncm_register.svg
>    :alt: Application of IPC peripheral registers.

Reviewer 3: ```suggestion
.. figure:: images/ncm_register.svg
   :alt: Application of IPC peripheral registers
   
   IPC peripheral registers
```

> The ``GPMEM[0]`` register is divided into two 16bit parts.
> This is shown in the figure above.

Reviewer 3: ```suggestion
The ``GPMEM[0]`` register is divided into two 16-bit parts.
```

> The value of the counter - *COUNTER* is incremented by the network core every :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC` Kconfig option in milliseconds.

Reviewer 3: ```suggestion
The ``COUNTER`` value is incremented by the network core every :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC`.
```

The application core periodically calls the :c:func:`ncm_net_status_check` function to check the status of the network core.

> If the network core is suspended, the counter value is not updated. The :c:func:`ncm_net_status_check` function returns the error code ``-EBUSY``.

Reviewer 3: ```suggestion
If the network core is suspended, the counter value is not updated.
The :c:func:`ncm_net_status_check` function returns error code ``-EBUSY``.
```

> The :c:func:`ncm_net_status_check` function should be called less frequently than the value of the :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC` Kconfig option.

Reviewer 3: ```suggestion
The :c:func:`ncm_net_status_check` function should be called less frequently than the value set in the :kconfig:option:`CONFIG_NCM_APP_FEEDING_INTERVAL_MSEC` Kconfig option.
```

Author: I added to this sentence why it should be.

> The *FLAGS* field contains the following entries:
> 
>    - Reset - Bit 0

Reviewer 3: ```suggestion
The ``FLAGS`` field contains entries ``Reset`` and ``Bit 0``.

```

Author: The point is that the FLAGS field contains one flag - Reset.
And it is bit zero of the FLAGS field.

Reviewer 3: Something like: "The ``FLAGS`` field has the ``Reset`` flag as the bit 0."?

Author: OK for me.

The reset bit is set at the boot of the network core.

> It informs the network core that a reset has occurred.

Author: Right.

> During startup of the networkore core, in addition to setting the reset bit, the cause of the reset is written to the IPC register ``GPMEM[1]``.

Reviewer 3: ```suggestion
During the startup of the network core, the reset bit is set and the cause of the reset is written to the IPC register ``GPMEM[1]``.
```

> This value is rewritten from the networkore core's ``RESET.RESETREAS`` register.

Reviewer 2: ```suggestion
This value is rewritten from the network core's ``RESET.RESETREAS`` register.
```

> See the documentation for a detailed description of the bits in this register: `nRF5340 Reset reason`_

Reviewer 3: ```suggestion
For a detailed description of the bits in this register, see `nRF5340 reset reason`_.
```

Reviewer 2: 'Reset reason' is a name of a sub-section under the 'RESETREAS' heading, and 'nRF5340' is not a part of it. I suggest using a section name, i.e., RESETREAS, as the exact 'Reset reason' sub-section is not linkable. What do you think Pekka? :)

```suggestion
For a detailed description of the bits in this register, see the `RESETREAS`_ section for nRF5340.
```

Reviewer 3: Agree, I did not check the link end :(

The :c:func:`ncm_net_status_check` function returns the following error codes:

>    - ``-EBUSY`` if the network core is suspended
>    - ``-EFAULT`` if a network core reset has occurred

Reviewer 3: ```suggestion
* ``-EBUSY`` if the network core is suspended
* ``-EFAULT`` if a network core reset has occurred
```

> The function takes as parameter a pointer to a variable of type uin32_t.

Reviewer 3: ```suggestion
The function takes a pointer to a variable of type uin32_t as a parameter.
```

> The function takes as parameter a pointer to a variable of type uin32_t.

Reviewer 4: Is it justified to describe the implementation in such details in this part of the documentation? It seems to me it can easily become outdated. It should be enough to document the API thoroughly in the relevant header.
Also, typo: uint32_t

Author: I added this description because I wanted to say that the function, in addition to the error codes, also returns the cause of network core reset.

Reviewer 4: OK, I'm fine with leaving it as it is.

When a network core reset is detected, the cause of the reset is stored in this pointer.

Configuration
*************

> To enable Network Core Monitor, set the :kconfig:option:`CONFIG_NET_CORE_MONITOR=y` Kconfig option on both network and application cores.

Reviewer 3: ```suggestion
To enable this library, set the :kconfig:option:`CONFIG_NET_CORE_MONITOR` Kconfig option to ``y`` on both network and application cores.
```

The :kconfig:option:`NCM_APP_FEEDING_INTERVAL_MSEC` Kconfig option specifies how often the counter is updated by the network core.
The default value is 500 milliseconds.

> The :kconfig:option:`CONFIG_RESET_INIT_PRIORITY` Kconfig option is a priority run function whose purpose is to read the cause of the processor reset and pass this information to the application core.

Reviewer 4: I don't understand this sentence, it states that the Kconfig option is a function. Please rephrase

Author: Improved.

> The :kconfig:option:`CONFIG_RESET_INIT_PRIORITY` Kconfig option is a priority run function whose purpose is to read the cause of the processor reset and pass this information to the application core.
> It is executed at the network core boot.

Reviewer 3: ```suggestion
The :kconfig:option:`CONFIG_RESET_INIT_PRIORITY` Kconfig option sets the priority run function.
The function reads the cause of the processor reset and passes this information to the application core.
It is executed at the network core boot.
```

Author: I changed the description of this parameter a bit.

Usage
*****

> Enable the Network Core Monitor module in the configuration file by setting the :kconfig:option:`CONFIG_NET_CORE_MONITOR` Kconfig option.

Reviewer 3: ```suggestion
To enable the Network core monitor library, set the :kconfig:option:`CONFIG_NET_CORE_MONITOR` Kconfig option.
```

On the application core, periodically call the :c:func:`ncm_net_status_check` function to monitor the state of the network core.
See the following usage example.

.. code-block::

   #include "net_core_monitor.h"
   ...
   static void print_reset(uint32_t reset_reas)
   {
      if (reset_reas & NRF_RESET_RESETREAS_RESETPIN_MASK) {
         printk("Reset by pin-reset\n");
      } else if (reset_reas & NRF_RESET_RESETREAS_DOG0_MASK) {
         printk("Reset by application watchdog timer 0 \n");
      } else if (reset_reas & NRF_RESET_RESETREAS_SREQ_MASK) {
         printk("Reset by soft-reset\n");
      } else if (reset_reas) {
         printk("Reset by a different source (0x%08X)\n", reset_reas);
      }
   }

   int main(void)
   {
      uint32_t reset_reas;
      ...
      for (;;) {
         ret = ncm_net_status_check(&reset_reas);
         if (ret == -EBUSY) {
            /* do something*/
         } else if (ret == -EFAULT) {
            print_reset(reset_reas);
         }
         k_sleep(K_MSEC(1000));
      }
   }

Dependencies
************

> The module uses two general purpose registers of the application core's IPC peripheral.

Reviewer 3: ```suggestion
The module uses two general-purpose registers, ``GPMEM[0]`` and ``GPMEM[1]``, of the application core's IPC peripheral.
```

> These are ``GPMEM[0]`` and ``GPMEM[1]``.

Reviewer 2: ```suggestion

```

API documentation
*****************

> 

Reviewer 2: redundant empty line
